File: examples/python/cart/cart_logic.py
"""Cart business logic - command handlers."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains
from state import CartState


class CommandRejectedError(Exception):
    """Command was rejected due to business rule violation."""


def handle_create_cart(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if state.exists():
        raise CommandRejectedError("Cart already exists")

    cmd = domains.CreateCart()
    command_any.Unpack(cmd)

    if not cmd.customer_id:
        raise CommandRejectedError("Customer ID is required")

    log.info("creating_cart", customer_id=cmd.customer_id)

    event = domains.CartCreated(
        customer_id=cmd.customer_id,
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_add_item(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")

    cmd = domains.AddItem()
    command_any.Unpack(cmd)

    if not cmd.product_id:
        raise CommandRejectedError("Product ID is required")
    if cmd.quantity <= 0:
        raise CommandRejectedError("Quantity must be positive")

    new_subtotal = state.subtotal_cents + (cmd.quantity * cmd.unit_price_cents)

    log.info("adding_item", product_id=cmd.product_id, quantity=cmd.quantity)

    event = domains.ItemAdded(
        product_id=cmd.product_id,
        name=cmd.name,
        quantity=cmd.quantity,
        unit_price_cents=cmd.unit_price_cents,
        new_subtotal=new_subtotal,
        added_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_update_quantity(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")

    cmd = domains.UpdateQuantity()
    command_any.Unpack(cmd)

    if cmd.product_id not in state.items:
        raise CommandRejectedError("Item not in cart")
    if cmd.new_quantity <= 0:
        raise CommandRejectedError("Quantity must be positive")

    item = state.items[cmd.product_id]
    old_subtotal = item.quantity * item.unit_price_cents
    new_item_subtotal = cmd.new_quantity * item.unit_price_cents
    new_subtotal = state.subtotal_cents - old_subtotal + new_item_subtotal

    log.info("updating_quantity", product_id=cmd.product_id, new_quantity=cmd.new_quantity)

    event = domains.QuantityUpdated(
        product_id=cmd.product_id,
        old_quantity=item.quantity,
        new_quantity=cmd.new_quantity,
        new_subtotal=new_subtotal,
        updated_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_remove_item(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")

    cmd = domains.RemoveItem()
    command_any.Unpack(cmd)

    if cmd.product_id not in state.items:
        raise CommandRejectedError("Item not in cart")

    item = state.items[cmd.product_id]
    item_subtotal = item.quantity * item.unit_price_cents
    new_subtotal = state.subtotal_cents - item_subtotal

    log.info("removing_item", product_id=cmd.product_id)

    event = domains.ItemRemoved(
        product_id=cmd.product_id,
        quantity=item.quantity,
        new_subtotal=new_subtotal,
        removed_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_apply_coupon(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")
    if state.coupon_code:
        raise CommandRejectedError("Coupon already applied")

    cmd = domains.ApplyCoupon()
    command_any.Unpack(cmd)

    if not cmd.code:
        raise CommandRejectedError("Coupon code is required")

    if cmd.coupon_type == "percentage":
        if cmd.value < 0 or cmd.value > 100:
            raise CommandRejectedError("Percentage must be 0-100")
        discount_cents = (state.subtotal_cents * cmd.value) // 100
    elif cmd.coupon_type == "fixed":
        if cmd.value < 0:
            raise CommandRejectedError("Fixed discount cannot be negative")
        discount_cents = min(cmd.value, state.subtotal_cents)
    else:
        raise CommandRejectedError("Invalid coupon type")

    log.info("applying_coupon", code=cmd.code, discount_cents=discount_cents)

    event = domains.CouponApplied(
        coupon_code=cmd.code,
        coupon_type=cmd.coupon_type,
        value=cmd.value,
        discount_cents=discount_cents,
        applied_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_clear_cart(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")

    log.info("clearing_cart")

    event = domains.CartCleared(
        new_subtotal=0,
        cleared_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_checkout(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")
    if not state.items:
        raise CommandRejectedError("Cart is empty")

    log.info("checking_out")

    event = domains.CartCheckedOut(
        final_subtotal=state.subtotal_cents,
        discount_cents=state.discount_cents,
        checked_out_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/ruby/cart/lib/cart_logic.rb
# frozen_string_literal: true

require 'time'
require_relative 'command_validation_error'

# Pure business logic for cart aggregate.
# No gRPC dependencies - can be tested in isolation.
module CartLogic
  CartItem = Struct.new(:sku, :name, :quantity, :unit_price_cents, keyword_init: true)

  CartState = Struct.new(:customer_id, :items, :coupon_code, :discount_cents, :status, keyword_init: true) do
    def initialize(customer_id: '', items: [], coupon_code: '', discount_cents: 0, status: :uninitialized)
      super
    end

    def exists?
      status != :uninitialized
    end

    def active?
      status == :active
    end

    def subtotal_cents
      items.sum { |item| item.quantity * item.unit_price_cents }
    end
  end

  class << self
    def rebuild_state(event_book)
      state = CartState.new

      return state if event_book.nil? || event_book.pages.empty?

      # Apply events
      event_book.pages.each do |page|
        next unless page.event

        state = apply_event(state, page.event)
      end

      state
    end

    def handle_create_cart(state, customer_id)
      raise CommandValidationError.failed_precondition('Cart already exists') if state.exists?
      raise CommandValidationError.invalid_argument('Customer ID is required') if customer_id.nil? || customer_id.empty?

      Examples::CartCreated.new(
        customer_id: customer_id,
        created_at: now_timestamp
      )
    end

    def handle_add_item(state, sku, name, quantity, unit_price_cents)
      raise CommandValidationError.failed_precondition('Cart does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Cart is not active') unless state.active?
      raise CommandValidationError.invalid_argument('SKU is required') if sku.nil? || sku.empty?
      raise CommandValidationError.invalid_argument('Quantity must be positive') if quantity <= 0

      Examples::ItemAdded.new(
        sku: sku,
        name: name || '',
        quantity: quantity,
        unit_price_cents: unit_price_cents
      )
    end

    def handle_update_quantity(state, sku, quantity)
      raise CommandValidationError.failed_precondition('Cart does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Cart is not active') unless state.active?
      raise CommandValidationError.invalid_argument('SKU is required') if sku.nil? || sku.empty?
      raise CommandValidationError.invalid_argument('Quantity must be positive') if quantity <= 0

      item = state.items.find { |i| i.sku == sku }
      raise CommandValidationError.failed_precondition("Item #{sku} not in cart") unless item

      Examples::QuantityUpdated.new(
        sku: sku,
        old_quantity: item.quantity,
        new_quantity: quantity
      )
    end

    def handle_remove_item(state, sku)
      raise CommandValidationError.failed_precondition('Cart does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Cart is not active') unless state.active?
      raise CommandValidationError.invalid_argument('SKU is required') if sku.nil? || sku.empty?

      item = state.items.find { |i| i.sku == sku }
      raise CommandValidationError.failed_precondition("Item #{sku} not in cart") unless item

      Examples::ItemRemoved.new(sku: sku)
    end

    def handle_apply_coupon(state, coupon_code, discount_cents)
      raise CommandValidationError.failed_precondition('Cart does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Cart is not active') unless state.active?
      raise CommandValidationError.invalid_argument('Coupon code is required') if coupon_code.nil? || coupon_code.empty?
      raise CommandValidationError.failed_precondition('Coupon already applied') unless state.coupon_code.empty?

      Examples::CouponApplied.new(
        coupon_code: coupon_code,
        discount_cents: discount_cents
      )
    end

    def handle_clear_cart(state)
      raise CommandValidationError.failed_precondition('Cart does not exist') unless state.exists?

      Examples::CartCleared.new(cleared_at: now_timestamp)
    end

    def handle_checkout(state, loyalty_points_to_use)
      raise CommandValidationError.failed_precondition('Cart does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Cart is not active') unless state.active?
      raise CommandValidationError.failed_precondition('Cart is empty') if state.items.empty?

      line_items = state.items.map do |item|
        Examples::LineItem.new(
          sku: item.sku,
          name: item.name,
          quantity: item.quantity,
          unit_price_cents: item.unit_price_cents
        )
      end

      Examples::CartCheckoutRequested.new(
        customer_id: state.customer_id,
        items: line_items,
        subtotal_cents: state.subtotal_cents,
        discount_cents: state.discount_cents,
        loyalty_points_to_use: loyalty_points_to_use || 0
      )
    end

    private

    def apply_event(state, event_any)
      type_url = event_any.type_url

      if type_url.end_with?('CartCreated')
        event = Examples::CartCreated.decode(event_any.value)
        CartState.new(
          customer_id: event.customer_id,
          items: [],
          coupon_code: '',
          discount_cents: 0,
          status: :active
        )
      elsif type_url.end_with?('ItemAdded')
        event = Examples::ItemAdded.decode(event_any.value)
        existing_item = state.items.find { |i| i.sku == event.sku }
        new_items = if existing_item
                      state.items.map do |i|
                        if i.sku == event.sku
                          CartItem.new(
                            sku: i.sku,
                            name: i.name,
                            quantity: i.quantity + event.quantity,
                            unit_price_cents: i.unit_price_cents
                          )
                        else
                          i
                        end
                      end
                    else
                      state.items + [CartItem.new(
                        sku: event.sku,
                        name: event.name,
                        quantity: event.quantity,
                        unit_price_cents: event.unit_price_cents
                      )]
                    end
        CartState.new(
          customer_id: state.customer_id,
          items: new_items,
          coupon_code: state.coupon_code,
          discount_cents: state.discount_cents,
          status: state.status
        )
      elsif type_url.end_with?('QuantityUpdated')
        event = Examples::QuantityUpdated.decode(event_any.value)
        new_items = state.items.map do |i|
          if i.sku == event.sku
            CartItem.new(
              sku: i.sku,
              name: i.name,
              quantity: event.new_quantity,
              unit_price_cents: i.unit_price_cents
            )
          else
            i
          end
        end
        CartState.new(
          customer_id: state.customer_id,
          items: new_items,
          coupon_code: state.coupon_code,
          discount_cents: state.discount_cents,
          status: state.status
        )
      elsif type_url.end_with?('ItemRemoved')
        event = Examples::ItemRemoved.decode(event_any.value)
        new_items = state.items.reject { |i| i.sku == event.sku }
        CartState.new(
          customer_id: state.customer_id,
          items: new_items,
          coupon_code: state.coupon_code,
          discount_cents: state.discount_cents,
          status: state.status
        )
      elsif type_url.end_with?('CouponApplied')
        event = Examples::CouponApplied.decode(event_any.value)
        CartState.new(
          customer_id: state.customer_id,
          items: state.items,
          coupon_code: event.coupon_code,
          discount_cents: event.discount_cents,
          status: state.status
        )
      elsif type_url.end_with?('CartCleared')
        CartState.new(
          customer_id: state.customer_id,
          items: [],
          coupon_code: '',
          discount_cents: 0,
          status: :cleared
        )
      else
        state
      end
    end

    def now_timestamp
      Google::Protobuf::Timestamp.new(seconds: Time.now.to_i)
    end
  end
end
